quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,71 +198,179 @@ impl Recv {
-    pub bufs: usize,
-    pub read: usize,
+    /// Read the next chunk from the stream
+    ///
+    /// Return the next chunk of data from the stream, if any are available. `max_length` may be
+    /// used to restrict the maximum size of the returned `Bytes` (but `usize::MAX` is supported,
+    /// too).
+    pub fn next(&mut self, max_length: usize) -> ReadResult {
+        if let Some(status) = &self.fused {
+            return ReadResult::from_status(*status);
+        }
+
+        let rs = self.recv.as_mut().unwrap();
+        let chunk = rs.assembler.read(max_length, self.ordered);
+        let status = match &chunk {
+            Some(chunk) => {
+                self.read += chunk.bytes.len() as u64;
+                ReadStatus::Readable
+            }
+            None => match rs.state {
+                RecvState::Recv { size } => {
+                    if size == Some(rs.end) && rs.assembler.bytes_read() == rs.end {
+                        self.streams.stream_freed(self.id, StreamHalf::Recv);
+                        self.fused = Some(ReadStatus::Finished);
+                        ReadStatus::Finished
+                    } else {
+                        self.streams.recv.insert(self.id, self.recv.take().unwrap());
+                        self.fused = Some(ReadStatus::Blocked);
+                        ReadStatus::Blocked
+                    }
+                }
+                RecvState::ResetRecvd { error_code, .. } => {
+                    rs.state = RecvState::Closed;
+                    self.streams.stream_freed(self.id, StreamHalf::Recv);
+                    self.fused = Some(ReadStatus::Reset(error_code));
+                    ReadStatus::Reset(error_code)
+                }
+                RecvState::Closed => unreachable!(),
+            },
+        };
+
+        let mut should_transmit = ShouldTransmit(false);
+        if status != ReadStatus::Readable {
+            should_transmit = self.finalize();
+        }
+
+        ReadResult {
+            chunk,
+            status,
+            should_transmit,
+        }
+    }
+
+    /// Finish reading, letting the caller know if the connection has something to transmit
+    pub fn finalize(&mut self) -> ShouldTransmit {
+        let mut rs = match self.recv.take() {
+            Some(rs) => rs,
+            None => return ShouldTransmit(false),
+        };
+
+        let (_, max_stream_data) = rs.max_stream_data(self.streams.stream_receive_window);
+        let max_data = self.streams.add_read_credits(self.read);
+        let max_dirty = self.streams.take_max_streams_dirty(self.id.dir());
+        self.pending
+            .post_read(self.id, max_data, max_stream_data, max_dirty);
+        max_data | max_stream_data | ShouldTransmit(max_dirty)
+    }
 }
 
-impl BytesRead for ReadChunks {
-    fn bytes_read(&self) -> u64 {
-        self.read as u64
+impl<'a> Drop for Chunks<'a> {
+    fn drop(&mut self) {
+        let _ = self.finalize();
```

Comment:
who executes the transmit if we are dropping the result?

Reply:
It has to be initiated by someone. To be fair, I don't think this is any worse than what we have today. And this would only happen if the caller did not read out `next()` until it contained `status != Readable`.

Reply:
Good idea, done.